File: 2023/typescript/src/day2/day2.ts
import { readAsStringArray } from "../utils/getInput";

type RGB = {
  [id: string]: number;
};

const validate = (
  input: string[],
  red: number,
  blue: number,
  green: number,
): [number[], number[]] => {
  const map: RGB = {
    red: red,
    blue: blue,
    green: green,
  };
  const validGames: number[] = [];
  const power: number[] = [];
  input.map((game) => {
    const [id, cubes] = game.replace("Game ", "").split(":");
    const minimum: RGB = {
      red: 0,
      blue: 0,
      green: 0,
    };
    const validGame = cubes
      .split(";")
      .map((sets) =>
        sets
          .split(",")
          .map((set) => {
            const [count, color] = set.trim().split(" ");
            if (parseInt(count) > minimum[color])
              minimum[color] = parseInt(count);
            return parseInt(count) <= map[color];
          })
          .reduce((acc, set) => acc && set),
      )
      .reduce((acc, sets) => acc && sets);
    if (validGame) validGames.push(parseInt(id));
    power.push(minimum.red * minimum.blue * minimum.green);
  });
  return [validGames, power];
};

export const partOne = (filename: string): number => {
  const input = readAsStringArray(filename);
  const [ids, _] = validate(input, 12, 14, 13);
  return ids.reduce((acc, id) => acc + id);
};

export const partTwo = (filename: string): number => {
  const input = readAsStringArray(filename);
  const [_, power] = validate(input, 12, 14, 13);
  return power.reduce((acc, pow) => acc + pow);
};

File: golang/2023/day2/day2.go
package day2

import (
	"aoc/golang/utils"
	"fmt"
	"strconv"
	"strings"
)

func validateGames(input []string) (int, int) {
	const maxRed int = 12
	const maxGreen int = 13
	const maxBlue int = 14

	score := 0
	power := 0
	for _, line := range input {
		minRed := 0
		minBlue := 0
		minGreen := 0
		validGame := true

		splitLine := (strings.Split(line, ":"))
		id, _ := strconv.Atoi(strings.TrimPrefix(splitLine[0], "Game "))
		sets := strings.Split(splitLine[1], ";")

		for _, set := range sets {
			colors := strings.Split(set, ",")

			for _, color := range colors {
				cubes := strings.Fields(color)
				cubeCount, _ := strconv.Atoi(cubes[0])
				switch cubes[1] {
				case "red":
					validGame = validGame && (cubeCount <= maxRed)
					if cubeCount > minRed {
						minRed = cubeCount
					}
				case "green":
					validGame = validGame && (cubeCount <= maxGreen)
					if cubeCount > minGreen {
						minGreen = cubeCount
					}
				case "blue":
					validGame = validGame && (cubeCount <= maxBlue)
					if cubeCount > minBlue {
						minBlue = cubeCount
					}
				}
			}
		}

		if validGame {
			score += id
		}
		power += minRed * minBlue * minGreen
	}
	return score, power
}

func runPartOne(filename string) int {
	fmt.Println("Part One")
	input := utils.ReadAsSlice[string](filename)
	result, _ := validateGames(input)
	return result
}

func runPartTwo(filename string) int {
	fmt.Println("Part Two")
	input := utils.ReadAsSlice[string](filename)
	_, result := validateGames(input)
	return result
}
